File: app/src/main/java/knf/kuma/animeinfo/ActivityAnime.java
package knf.kuma.animeinfo;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.InflateException;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.ContentViewEvent;
import com.crashlytics.android.answers.ShareEvent;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityOptionsCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import es.munix.multidisplaycast.CastManager;
import knf.kuma.R;
import knf.kuma.animeinfo.viewholders.AnimeActivityHolder;
import knf.kuma.commons.EAHelper;
import knf.kuma.commons.PatternUtil;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.FavsDAO;
import knf.kuma.database.dao.SeeingDAO;
import knf.kuma.pojos.AnimeObject;
import knf.kuma.pojos.ExplorerObject;
import knf.kuma.pojos.FavoriteObject;
import knf.kuma.pojos.NotificationObj;
import knf.kuma.pojos.QueueObject;
import knf.kuma.pojos.RecentObject;
import knf.kuma.pojos.RecordObject;
import knf.kuma.pojos.SeeingObject;
import knf.kuma.recommended.RankType;
import knf.kuma.recommended.RecommendHelper;
import knf.kuma.widgets.emision.WEListItem;
import xdroid.toaster.Toaster;

public class ActivityAnime extends AppCompatActivity implements AnimeActivityHolder.Interface {
    public static int REQUEST_CODE = 558;
    private boolean isEdited = false;
    private AnimeViewModel viewModel;
    private AnimeActivityHolder holder;
    private FavoriteObject favoriteObject;
    private FavsDAO dao = CacheDB.INSTANCE.favsDAO();
    private SeeingDAO seeingDAO = CacheDB.INSTANCE.seeingDAO();
    private List<AnimeObject.WebInfo.AnimeChapter> chapters = new ArrayList<>();
    private List<String> genres = new ArrayList<>();

    public static void open(Fragment fragment, RecentObject object, View view, int position) {
        Intent intent = new Intent(fragment.getContext(), ActivityAnime.class);
        intent.setData(Uri.parse(object.anime));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        intent.putExtra("position", position);
        fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(fragment.getActivity(), view, "img").toBundle());
    }

    public static void open(Fragment fragment, AnimeObject object, View view) {
        open(fragment, object, view, true, true);
    }

    public static void open(Fragment fragment, AnimeObject object, View view, boolean persist, boolean animate) {
        Intent intent = new Intent(fragment.getContext(), ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        intent.putExtra("persist", persist);
        intent.putExtra("noTransition", !animate);
        fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(fragment.getActivity(), view, "img").toBundle());
    }

    public static void open(Activity activity, AnimeObject object, View view, boolean persist, boolean animate) {
        Intent intent = new Intent(activity, ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        intent.putExtra("persist", persist);
        intent.putExtra("noTransition", !animate);
        activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, "img").toBundle());
    }

    public static void open(Fragment fragment, ExplorerObject object, View view) {
        Intent intent = new Intent(fragment.getContext(), ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", String.valueOf(object.key));
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(fragment.getActivity(), view, "img").toBundle());
    }

    public static void open(Activity activity, RecordObject object, View view) {
        Intent intent = new Intent(activity, ActivityAnime.class);
        intent.setData(Uri.parse(object.animeObject.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.animeObject.aid));
        intent.putExtra("persist", true);
        intent.putExtra("isRecord", true);
        activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, "img").toBundle());
    }

    public static void open(Activity activity, SeeingObject object, View view) {
        Intent intent = new Intent(activity, ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.title);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        intent.putExtra("persist", true);
        intent.putExtra("noTransition", true);
        intent.putExtra("isRecord", true);
        activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, "img").toBundle());
    }

    public static void open(Context context, AnimeObject object) {
        Intent intent = new Intent(context, ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        context.startActivity(intent);
    }

    public static void open(Fragment fragment, FavoriteObject object, View view) {
        Intent intent = new Intent(fragment.getContext(), ActivityAnime.class);
        intent.setData(Uri.parse(object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        intent.putExtra("img", PatternUtil.getCover(object.aid));
        intent.putExtra("from_fav", true);
        fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(fragment.getActivity(), view, "img").toBundle());
    }

    public static void open(Activity activity, QueueObject object, String img, View view) {
        Intent intent = new Intent(activity, ActivityAnime.class);
        intent.putExtra("title", object.chapter.name);
        intent.putExtra("aid", object.chapter.aid);
        intent.putExtra("img", PatternUtil.getCover(object.chapter.aid));
        intent.putExtra("aid_only", true);
        activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, "img").toBundle());
    }

    public static void open(Fragment fragment, AnimeObject.WebInfo.AnimeRelated object) {
        Intent intent = new Intent(fragment.getContext(), ActivityAnime.class);
        intent.setData(Uri.parse("https://animeflv.net/" + object.link));
        intent.putExtra("title", object.name);
        intent.putExtra("aid", object.aid);
        fragment.startActivityForResult(intent, REQUEST_CODE);
    }

    public static Intent getSimpleIntent(Context context, WEListItem item) {
        Intent intent = new Intent(context, ActivityAnime.class);
        intent.setData(Uri.parse(item.link));
        intent.putExtra("title", item.title);
        intent.putExtra("aid", item.aid);
        intent.putExtra("img", PatternUtil.getCover(item.aid));
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(EAHelper.getThemeNA(this));
        super.onCreate(savedInstanceState);
        try {
            setContentView(R.layout.activity_anime_info);
        } catch (InflateException e) {
            setContentView(R.layout.activity_anime_info_nwv);
        }
        if (getIntent().getBooleanExtra("notification", false))
            sendBroadcast(NotificationObj.fromIntent(getIntent()).getBroadcast(this));
        holder = new AnimeActivityHolder(this);
        setSupportActionBar(holder.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(false);
        holder.toolbar.setNavigationOnClickListener(view -> closeActivity());
        viewModel = ViewModelProviders.of(this).get(AnimeViewModel.class);
        if (getIntent().getBooleanExtra("aid_only", false))
            viewModel.init(getIntent().getStringExtra("aid"));
        else
            viewModel.init(this, getIntent().getDataString(), getIntent().getBooleanExtra("persist", true));
        load();
    }

    private void load() {
        viewModel.getLiveData().observe(this, object -> {
            if (object != null) {
                Crashlytics.setString("screen", "Anime: " + object.name);
                Answers.getInstance().logContentView(new ContentViewEvent().putContentName(object.name).putContentType(object.type).putContentId(object.aid));
                chapters = object.chapters;
                genres = object.genres;
                favoriteObject = new FavoriteObject(object);
                holder.setTitle(object.name);
                holder.loadImg(PatternUtil.getCover(object.aid), v -> startActivity(new Intent(ActivityAnime.this, ActivityImgFull.class).setData(Uri.parse(PatternUtil.getCover(object.aid))).putExtra("title", object.name), ActivityOptionsCompat.makeSceneTransitionAnimation(ActivityAnime.this, holder.imageView, "img").toBundle()));
                holder.setFABState(dao.isFav(favoriteObject.key), seeingDAO.getByAid(favoriteObject.aid) != null);
                holder.showFAB();
                supportInvalidateOptionsMenu();
                RecommendHelper.registerAll(genres, RankType.CHECK);
            } else {
                Toaster.toast("Error al cargar informacion del anime");
                onBackPressed();
            }
        });
    }

    private void setResult() {
        isEdited = true;
    }

    @Override
    public void onFabClicked(FloatingActionButton actionButton) {
        setResult();
        boolean isfav = dao.isFav(favoriteObject.key);
        boolean isSeeing = seeingDAO.getByAid(favoriteObject.aid) != null;
        if (isfav && isSeeing)
            onFabLongClicked(actionButton);
        else if (isfav) {
            holder.setFABState(false);
            dao.deleteFav(favoriteObject);
            RecommendHelper.registerAll(genres, RankType.UNFAV);
        } else if (isSeeing) {
            onFabLongClicked(actionButton);
        } else {
            holder.setFABState(true);
            dao.addFav(favoriteObject);
            RecommendHelper.registerAll(genres, RankType.FAV);
        }
    }

    @Override
    public void onFabLongClicked(FloatingActionButton actionButton) {
        try {
            final SeeingObject seeingObject = seeingDAO.getByAid(favoriteObject.aid);
            boolean isfav = dao.isFav(favoriteObject.key);
            boolean isSeeing = seeingObject != null;
            if (isSeeing) {
                new MaterialDialog.Builder(this)
                        .content("¿Convertir en favorito?")
                        .positiveText("si")
                        .negativeText("no")
                        .neutralText("dropear")
                        .onPositive((dialog, which) -> {
                            dao.addFav(favoriteObject);
                            seeingDAO.remove(seeingObject);
                            holder.setFABState(true);
                            RecommendHelper.registerAll(genres, RankType.FAV);
                        })
                        .onNegative((dialog, which) -> {
                            dao.deleteFav(favoriteObject);
                            holder.setFABSeeing();
                            RecommendHelper.registerAll(genres, RankType.UNFAV);
                        })
                        .onNeutral((dialog, which) -> {
                            seeingDAO.remove(seeingObject);
                            holder.setFABState(isfav);
                            RecommendHelper.registerAll(genres, RankType.UNFOLLOW);
                        }).build().show();
            } else if (!isfav) {
                new MaterialDialog.Builder(this)
                        .items("Seguir", "Seguir y favorito")
                        .itemsCallbackSingleChoice(0, (dialog, itemView, which, text) -> {
                            switch (which) {
                                case 0:
                                    seeingDAO.add(SeeingObject.fromAnime(favoriteObject));
                                    holder.setFABSeeing();
                                    RecommendHelper.registerAll(genres, RankType.FOLLOW);
                                    Toaster.toast("Agregado a animes seguidos");
                                    break;
                                case 1:
                                    dao.addFav(favoriteObject);
                                    seeingDAO.add(SeeingObject.fromAnime(favoriteObject));
                                    holder.setFABState(true, true);
                                    RecommendHelper.registerAll(genres, RankType.FAV);
                                    RecommendHelper.registerAll(genres, RankType.FOLLOW);
                                    Toaster.toast("Agregado a animes seguidos y favoritos");
                                    break;
                            }
                            return true;
                        })
                        .positiveText("aceptar")
                        .negativeText("cancelar")
                        .build().show();
            } else {
                seeingDAO.add(SeeingObject.fromAnime(favoriteObject));
                holder.setFABState(true, true);
                RecommendHelper.registerAll(genres, RankType.FOLLOW);
                Toaster.toast("Agregado a animes seguidos");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onImgClicked(ImageView imageView) {

    }

    @Override
    public void onNeedRecreate() {
        try {
            if (!getIntent().getBooleanExtra("aid_only", false)) {
                viewModel.reload(this, getIntent().getDataString(), getIntent().getBooleanExtra("persist", true));
                load();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (favoriteObject != null) {
            getMenuInflater().inflate(R.menu.menu_anime_info, menu);
            CastManager.getInstance().registerForActivity(this, menu, R.id.castMenu);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                share();
                break;
            case R.id.action_comments:
                new CommentsDialog(chapters).show(this);
                break;
        }
        return true;
    }

    private void share() {
        try {
            startActivity(Intent.createChooser(new Intent(Intent.ACTION_SEND)
                    .setType("text/plain")
                    .putExtra(Intent.EXTRA_TEXT, favoriteObject.name + "\n" + favoriteObject.link), "Compartir"));
            Answers.getInstance().logShare(new ShareEvent().putContentName(favoriteObject.name).putContentId(favoriteObject.aid));
        } catch (ActivityNotFoundException e) {
            Toaster.toast("No se encontraron aplicaciones para enviar");
        }
    }

    private void closeActivity() {
        holder.hideFABForce();
        if (getIntent().getBooleanExtra("from_fav", false) && isEdited) {
            finish();
        } else if (getIntent().getBooleanExtra("noTransition", false)) {
            finish();
        } else {
            supportFinishAfterTransition();
        }
    }

    @Override
    public void onBackPressed() {
        closeActivity();
    }
}

File: app/src/main/java/knf/kuma/animeinfo/ActivityAnime.kt
package knf.kuma.animeinfo

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.InflateException
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.activity.viewModels
import androidx.core.app.ActivityOptionsCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import com.google.android.material.floatingactionbutton.FloatingActionButton
import knf.kuma.R
import knf.kuma.achievements.AchievementManager
import knf.kuma.ads.showRandomInterstitial
import knf.kuma.animeinfo.img.ActivityImgFull
import knf.kuma.animeinfo.viewholders.AnimeActivityHolder
import knf.kuma.backup.firestore.syncData
import knf.kuma.commons.CastUtil
import knf.kuma.commons.DesignUtils
import knf.kuma.commons.EAHelper
import knf.kuma.commons.PatternUtil
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.doOnUI
import knf.kuma.custom.GenericActivity
import knf.kuma.database.CacheDB
import knf.kuma.directory.DirObject
import knf.kuma.directory.DirObjectCompact
import knf.kuma.pojos.AnimeObject
import knf.kuma.pojos.ExplorerObject
import knf.kuma.pojos.FavoriteObject
import knf.kuma.pojos.NotificationObj
import knf.kuma.pojos.QueueObject
import knf.kuma.pojos.RecentObject
import knf.kuma.pojos.RecordObject
import knf.kuma.pojos.SeeingObject
import knf.kuma.recommended.AnimeShortObject
import knf.kuma.recommended.RankType
import knf.kuma.recommended.RecommendHelper
import knf.kuma.search.SearchObject
import knf.kuma.search.SearchObjectFav
import knf.kuma.widgets.emision.WEListItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.sdk27.coroutines.onLongClick
import org.jetbrains.anko.toast
import xdroid.toaster.Toaster
import java.util.Locale
import kotlin.random.Random

class ActivityAnime : GenericActivity(), AnimeActivityHolder.Interface {
    private var isEdited = false
    private val viewModel: AnimeViewModel by viewModels()
    private val holder: AnimeActivityHolder by lazy { AnimeActivityHolder(this) }
    private var favoriteObject: FavoriteObject? = null
    private val dao = CacheDB.INSTANCE.favsDAO()
    private var chapters: MutableList<AnimeObject.WebInfo.AnimeChapter> = ArrayList()
    private var genres: MutableList<String> = ArrayList()
    private val aidOnly get() = intent?.getBooleanExtra(keyAidOnly, false) ?: false

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(EAHelper.getThemeNA())
        super.onCreate(savedInstanceState)
        try {
            setContentView(R.layout.activity_anime_info)
        } catch (e: InflateException) {
            setContentView(R.layout.activity_anime_info_nwv)
        }
        setSupportActionBar(holder.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(false)
        holder.toolbar.setNavigationOnClickListener { closeActivity() }
        if (aidOnly)
            viewModel.init(intent.getStringExtra(keyAid))
        else
            viewModel.init(this@ActivityAnime, intent.dataString, intent.getBooleanExtra(keyPersist, true))
        if (intent.getBooleanExtra(keyNotification, false))
            sendBroadcast(NotificationObj.fromIntent(intent).getBroadcast(this@ActivityAnime))
        load()
        checkBypass()
        showRandomInterstitial(this)
    }

    private fun load() {
        viewModel.liveData.observe(this, Observer { animeObject ->
            if (animeObject != null) {
                doOnUI {
                    chapters = animeObject.chapters ?: mutableListOf()
                    genres = animeObject.genres ?: mutableListOf()
                    if (PrefsUtil.isFamilyFriendly && genres.map { it.lowercase(Locale.getDefault()) }.contains("ecchi")) {
                        toast("Anime no familiar")
                        onBackPressed()
                    }
                    favoriteObject = FavoriteObject(animeObject)
                    favoriteObject?.let { fav ->
                        holder.imageView.onLongClick(returnValue = true) {
                            doAsync {
                                val isFav = dao.isFav(fav.key)
                                if (isFav) {
                                    holder.setFABState(false)
                                    dao.deleteFav(fav)
                                    RecommendHelper.registerAll(genres, RankType.UNFAV)
                                    doOnUI { toast("Removido de favoritos") }
                                } else {
                                    holder.setFABState(true)
                                    dao.addFav(fav)
                                    RecommendHelper.registerAll(genres, RankType.FAV)
                                    AchievementManager.onFavAdded(fav)
                                    doOnUI { toast("Añadido a favoritos") }
                                }
                                syncData { favs() }
                            }
                        }
                        dao.isFavLive(fav.key).observe(this, Observer { holder.setFABState(it) })
                    }
                    holder.setTitle(animeObject.name)
                    holder.loadImg(PatternUtil.getCover(animeObject.aid), View.OnClickListener {
                        startActivity(
                                Intent(this@ActivityAnime, ActivityImgFull::class.java)
                                        .setData(Uri.parse(PatternUtil.getCover(animeObject.aid)))
                                        .putExtra(keyTitle, animeObject.name), ActivityOptionsCompat.makeSceneTransitionAnimation(this@ActivityAnime, holder.imageView, "img")
                                .toBundle()
                        )
                    })
                    lifecycleScope.launch(Dispatchers.Main){
                        holder.setFABState(withContext(Dispatchers.IO) { dao.isFav(favoriteObject?.key ?: 0) })
                        holder.showFAB()
                    }
                    invalidateOptionsMenu()
                    RecommendHelper.registerAll(genres, RankType.CHECK)
                }
            } else {
                Toaster.toast("Error al cargar información del anime")
                onBackPressed()
            }
        })
    }

    private fun setResult() {
        isEdited = true
    }

    override fun onFabClicked(actionButton: FloatingActionButton) {
        lifecycleScope.launch(Dispatchers.Main) {
            setResult()
            favoriteObject?.let {
                val isFav = withContext(Dispatchers.IO) { dao.isFav(it.key) }
                if (isFav) {
                    holder.setFABState(false)
                    withContext(Dispatchers.IO) { dao.deleteFav(it) }
                    RecommendHelper.registerAll(genres, RankType.UNFAV)
                } else {
                    holder.setFABState(true)
                    withContext(Dispatchers.IO) { dao.addFav(it) }
                    RecommendHelper.registerAll(genres, RankType.FAV)
                    AchievementManager.onFavAdded(it)
                }
                syncData { favs() }
            }
        }
    }

    override fun onImgClicked(imageView: ImageView) {

    }

    override fun onBypassUpdated() {
        try {
            if (!aidOnly)
                viewModel.reload(this, intent.dataString, intent.getBooleanExtra(keyPersist, true))
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun getSnackbarAnchor(): View? {
        return findViewById(R.id.coordinator)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (favoriteObject != null) {
            menuInflater.inflate(R.menu.menu_anime_info, menu)
            CastUtil.registerActivity(this, menu, R.id.castMenu)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_share -> share()
        }
        return true
    }

    private fun share() {
        try {
            startActivity(Intent.createChooser(Intent(Intent.ACTION_SEND)
                    .setType("text/plain")
                    .putExtra(Intent.EXTRA_TEXT, favoriteObject?.name + "\n" + favoriteObject?.link), "Compartir"))
            AchievementManager.onShare()
        } catch (e: ActivityNotFoundException) {
            Toaster.toast("No se encontraron aplicaciones para enviar")
        }

    }

    private fun closeActivity() {
        holder.hideFABForce()
        if (intent.getBooleanExtra(keyFromFav, false) && isEdited) {
            finish()
        } else if (intent.getBooleanExtra(keyNoTransition, false)) {
            finish()
        } else {
            supportFinishAfterTransition()
        }
    }

    override fun onBackPressed() {
        closeActivity()
    }

    companion object {
        private var REQUEST_CODE = 558
        private const val keyTitle = "title"
        private const val keyAid = "aid"
        private const val keyImg = "img"
        private const val keyPosition = "persist"
        private const val keyPersist = "persist"
        private const val keyNoTransition = "noTransition"
        private const val keyIsRecord = "isRecord"
        private const val keyFromFav = "from_fav"
        private const val keyAidOnly = "aid_only"
        private const val keyNotification = "notification"
        private const val sharedImg = "img"

        fun open(fragment: Fragment, animeObject: SearchObject, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, animeObject: SearchObjectFav, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, recentObject: RecentObject, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(recentObject.url)
            intent.putExtra(keyTitle, recentObject.name)
            intent.putExtra(keyAid, recentObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(recentObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, animeObject: DirObjectCompact, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, animeObject: DirObject, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            try {
                val intent = Intent(fragment.context, DesignUtils.infoClass)
                intent.data = Uri.parse(animeObject.link)
                intent.putExtra(keyTitle, animeObject.name)
                intent.putExtra(keyAid, animeObject.aid)
                intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
                intent.putExtra(keyPersist, persist)
                intent.putExtra(keyNoTransition, !animate)
                fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        fun open(activity: Activity, animeObject: SearchObject, view: ImageView, persist: Boolean, animate: Boolean) {
            val intent = Intent(activity, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(activity: Activity?, animeObject: AnimeShortObject, view: ImageView, persist: Boolean, animate: Boolean) {
            val intent = Intent(activity, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            activity?.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, explorerObject: ExplorerObject, view: ImageView) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(explorerObject.link)
            intent.putExtra(keyTitle, explorerObject.name)
            intent.putExtra(keyAid, explorerObject.key.toString())
            intent.putExtra(keyImg, PatternUtil.getCover(explorerObject.aid))
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(activity: Activity, recordObject: RecordObject, view: ImageView) {
            val intent = Intent(activity, DesignUtils.infoClass)
            intent.data = Uri.parse(recordObject.animeObject.link)
            intent.putExtra(keyTitle, recordObject.name)
            intent.putExtra(keyAid, recordObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(recordObject.animeObject.aid))
            intent.putExtra(keyPersist, true)
            intent.putExtra(keyIsRecord, true)
            activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(activity: Activity?, seeingObject: SeeingObject) {
            activity ?: return
            val intent = Intent(activity, DesignUtils.infoClass)
            intent.data = Uri.parse(seeingObject.link)
            intent.putExtra(keyTitle, seeingObject.title)
            //intent.putExtra(keyAid, seeingObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(seeingObject.aid))
            intent.putExtra(keyPersist, true)
            intent.putExtra(keyNoTransition, true)
            intent.putExtra(keyIsRecord, true)
            activity.startActivity(intent)
        }

        fun open(context: Context, animeObject: SearchObject) {
            val intent = Intent(context, DesignUtils.infoClass)
            intent.data = Uri.parse(animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            context.startActivity(intent)
        }

        fun open(fragment: Fragment, favoriteObject: FavoriteObject, view: ImageView) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse(favoriteObject.link)
            intent.putExtra(keyTitle, favoriteObject.name)
            intent.putExtra(keyAid, favoriteObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(favoriteObject.aid))
            intent.putExtra(keyFromFav, true)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(activity: Activity, queueObject: QueueObject, view: ImageView) {
            val intent = Intent(activity, DesignUtils.infoClass)
            intent.putExtra(keyTitle, queueObject.chapter.name)
            intent.putExtra(keyAid, queueObject.chapter.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(queueObject.chapter.aid))
            intent.putExtra(keyAidOnly, true)
            activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(fragment: Fragment, animeRelated: AnimeObject.WebInfo.AnimeRelated) {
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse("https://www3.animeflv.net/" + animeRelated.link)
            intent.putExtra(keyTitle, animeRelated.name)
            intent.putExtra(keyAid, animeRelated.aid)
            fragment.startActivityForResult(intent, REQUEST_CODE)
        }

        fun open(fragment: Fragment, animeObject: AnimeObject.WebInfo.AnimeRelated, view: ImageView, persist: Boolean = true, animate: Boolean = true) {
            val activity = fragment.activity ?: return
            val intent = Intent(fragment.context, DesignUtils.infoClass)
            intent.data = Uri.parse("https://www3.animeflv.net/" + animeObject.link)
            intent.putExtra(keyTitle, animeObject.name)
            intent.putExtra(keyAid, animeObject.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(animeObject.aid))
            intent.putExtra(keyPersist, persist)
            intent.putExtra(keyNoTransition, !animate)
            fragment.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, view, sharedImg).toBundle())
        }

        fun open(context: Context, url: String) {
            val intent = Intent(context, DesignUtils.infoClass)
            intent.data = Uri.parse(url)
            context.startActivity(intent)
        }

        fun getSimpleIntent(context: Context, item: WEListItem): Intent {
            val intent = Intent(context, DesignUtils.infoClass)
            intent.data = Uri.parse(item.link)
            intent.action = "${Random.nextInt(1, 9000)}"
            intent.putExtra(keyTitle, item.title)
            intent.putExtra(keyAid, item.aid)
            intent.putExtra(keyImg, PatternUtil.getCover(item.aid))
            return intent
        }
    }
}
